File: typescript/src/command/command.ts
import { Session } from "../session";
import {
  CommandResult,
} from "../types/api-response";


/**
 * Handles command execution operations in the AgentBay cloud environment.
 */
export class Command {
  private session: Session;

  /**
   * Initialize a Command object.
   *
   * @param session - The Session instance that this Command belongs to.
   */
  constructor(session: Session) {
    this.session = session;
  }

  /**
   * Sanitizes error messages to remove sensitive information like API keys.
   *
   * @param error - The error to sanitize
   * @returns The sanitized error
   */
  private sanitizeError(error: any): any {
    if (!error) {
      return error;
    }

    const errorString = String(error);
    return errorString.replace(/Bearer\s+[^\s]+/g, "Bearer [REDACTED]");
  }

  /**
   * Execute a shell command with optional working directory and environment variables.
   *
   * Executes a shell command in the session environment with configurable timeout,
   * working directory, and environment variables. The command runs with session
   * user permissions in a Linux shell environment.
   *
   * @param command - The shell command to execute
   * @param timeoutMs - Timeout in milliseconds (default: 1000ms/1s). Maximum allowed
   *                    timeout is 50000ms (50s). If a larger value is provided,
   *                    it will be automatically limited to 50000ms
   * @param cwd - The working directory for command execution. If not specified,
   *              the command runs in the default session directory
   * @param envs - Environment variables as a dictionary of key-value pairs.
   *              These variables are set for the command execution only
   *
   * @returns Promise resolving to CommandResult containing:
   *          - success: Whether the command executed successfully (exitCode === 0)
   *          - output: Command output for backward compatibility (stdout + stderr)
   *          - exitCode: The exit code of the command execution (0 for success)
   *          - stdout: Standard output from the command execution
   *          - stderr: Standard error from the command execution
   *          - traceId: Trace ID for error tracking (only present when exitCode !== 0)
   *          - requestId: Unique identifier for this API request
   *          - errorMessage: Error description if execution failed
   *
   * @example
   * ```typescript
   * const agentBay = new AgentBay({ apiKey: 'your_api_key' });
   * const result = await agentBay.create();
   * if (result.success) {
   *   const cmdResult = await result.session.command.executeCommand('echo "Hello"', 5000);
   *   console.log('Command output:', cmdResult.output);
   *   console.log('Exit code:', cmdResult.exitCode);
   *   console.log('Stdout:', cmdResult.stdout);
   *   await result.session.delete();
   * }
   * ```
   *
   * @example
   * ```typescript
   * const agentBay = new AgentBay({ apiKey: 'your_api_key' });
   * const result = await agentBay.create();
   * if (result.success) {
   *   const cmdResult = await result.session.command.executeCommand(
   *     'pwd',
   *     5000,
   *     '/tmp',
   *     { TEST_VAR: 'test_value' }
   *   );
   *   console.log('Working directory:', cmdResult.stdout);
   *   await result.session.delete();
   * }
   * ```
   */
  async executeCommand(
    command: string,
    timeoutMs = 1000,
    cwd?: string,
    envs?: Record<string, string>
  ): Promise<CommandResult> {
    // Validate environment variables - strict type checking (before try block to allow Error to propagate)
    if (envs !== undefined) {
      const invalidVars: string[] = [];
      for (const [key, value] of Object.entries(envs)) {
        if (typeof key !== "string") {
          invalidVars.push(`key '${key}' (type: ${typeof key})`);
        }
        if (typeof value !== "string") {
          invalidVars.push(`value for key '${key}' (type: ${typeof value})`);
        }
      }
      
      if (invalidVars.length > 0) {
        throw new Error(
          `Invalid environment variables: all keys and values must be strings. ` +
          `Found invalid entries: ${invalidVars.join(", ")}`
        );
      }
    }

    try {
      // Limit timeout to maximum 50s (50000ms) as per SDK constraints
      const MAX_TIMEOUT_MS = 50000;
      if (timeoutMs > MAX_TIMEOUT_MS) {
        // Log warning (in production, you might want to use a proper logger)
        // Note: Warning is silently applied to maintain compatibility
        timeoutMs = MAX_TIMEOUT_MS;
      }

      // Build request arguments
      const args: Record<string, any> = {
        command,
        timeout_ms: timeoutMs,
      };
      if (cwd !== undefined) {
        args.cwd = cwd;
      }
      if (envs !== undefined) {
        args.envs = envs;
      }

      const result = await this.session.callMcpTool(
        "shell",
        args,
        false
      );

      if (result.success) {
        // Try to parse the new JSON format response
        try {
          // Parse JSON string from result.data
          let dataJson: any;
          if (typeof result.data === "string") {
            dataJson = JSON.parse(result.data);
          } else {
            dataJson = result.data;
          }

          // Extract fields from new format
          const stdout = dataJson.stdout || "";
          const stderr = dataJson.stderr || "";
          const exitCode = dataJson.exit_code || 0;
          const traceId = dataJson.traceId || "";

          // Determine success based on exit_code (0 means success)
          const success = exitCode === 0;

          // For backward compatibility, output should be stdout + stderr
          const output = stdout + stderr;

          return {
            requestId: result.requestId,
            success,
            output,
            exitCode: exitCode,
            stdout,
            stderr,
            traceId: traceId || undefined,
            errorMessage: result.errorMessage,
          };
        } catch (parseError) {
          // Fallback to old format if JSON parsing fails
          return {
            requestId: result.requestId,
            success: true,
            output: typeof result.data === "string" ? result.data : String(result.data),
            errorMessage: result.errorMessage,
          };
        }
      } else {
        // Try to parse error message as JSON (in case backend returns JSON in error)
        try {
          const errorData = typeof result.errorMessage === "string" 
            ? JSON.parse(result.errorMessage) 
            : result.errorMessage;
          
          if (errorData && typeof errorData === "object") {
            const stdout = errorData.stdout || "";
            const stderr = errorData.stderr || "";
            // Backend may return either "exit_code" or "errorCode", support both
            const exitCode = errorData.exit_code || errorData.errorCode || 0;
            const traceId = errorData.traceId || "";
            // For backward compatibility, output should be stdout + stderr
            const output = stdout + stderr;

            return {
              requestId: result.requestId,
              success: false,
              output,
              exitCode: exitCode,
              stdout,
              stderr,
              traceId: traceId || undefined,
              errorMessage: stderr || result.errorMessage,
            };
          }
        } catch {
          // If parsing fails, use original error message
        }

        return {
          requestId: result.requestId,
          success: false,
          output: "",
          errorMessage: result.errorMessage || "Failed to execute command",
        };
      }
    } catch (error) {
      return {
        requestId: "",
        success: false,
        output: "",
        errorMessage: `Failed to execute command: ${error}`,
      };
    }
  }

  /**
   * Alias of executeCommand().
   */
  async run(
    command: string,
    timeoutMs = 1000,
    cwd?: string,
    envs?: Record<string, string>
  ): Promise<CommandResult> {
    return await this.executeCommand(command, timeoutMs, cwd, envs);
  }

  /**
   * Alias of executeCommand().
   */
  async exec(
    command: string,
    timeoutMs = 1000,
    cwd?: string,
    envs?: Record<string, string>
  ): Promise<CommandResult> {
    return await this.executeCommand(command, timeoutMs, cwd, envs);
  }
}

File: golang/pkg/agentbay/command/command.go
package command

import (
	"encoding/json"
	"fmt"

	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// CommandResult represents the result of a command execution
type CommandResult struct {
	// Embed the basic API response structure
	models.ApiResponse
	// Success indicates whether the command execution was successful
	Success bool `json:"success"`
	// Output contains the command execution output (for backward compatibility, equals stdout + stderr)
	Output string `json:"output"`
	// ErrorMessage contains error message if the operation failed
	ErrorMessage string `json:"error_message,omitempty"`
	// ExitCode is the exit code of the command execution. Default is 0.
	ExitCode int `json:"exit_code"`
	// Stdout is the standard output from the command execution
	Stdout string `json:"stdout"`
	// Stderr is the standard error from the command execution
	Stderr string `json:"stderr"`
	// TraceID is the trace ID for error tracking. Only present when exit_code != 0. Used for quick problem localization.
	TraceID string `json:"trace_id,omitempty"`
}

// Command handles command execution operations in the AgentBay cloud environment.
type Command struct {
	Session interface {
		GetAPIKey() string
		GetClient() *mcp.Client
		GetSessionId() string
		IsVpc() bool
		NetworkInterfaceIp() string
		HttpPort() string
		FindServerForTool(toolName string) string
		CallMcpTool(toolName string, args interface{}, autoGenSession ...bool) (*models.McpToolResult, error)
	}
}

// NewCommand creates a new Command instance
func NewCommand(session interface {
	GetAPIKey() string
	GetClient() *mcp.Client
	GetSessionId() string
	IsVpc() bool
	NetworkInterfaceIp() string
	HttpPort() string
	FindServerForTool(toolName string) string
	CallMcpTool(toolName string, args interface{}, autoGenSession ...bool) (*models.McpToolResult, error)
}) *Command {
	return &Command{
		Session: session,
	}
}

// ==================== Option Definitions ====================

// commandOptions holds the configuration for command execution
type commandOptions struct {
	timeoutMs int
	cwd       string
	envs      map[string]string
}

// CommandOption is a function type for configuring ExecuteCommand options.
// This enables the Functional Options pattern for flexible and extensible API design.
type CommandOption func(*commandOptions)

// WithTimeoutMs sets the timeout for command execution in milliseconds.
// Maximum allowed timeout is 50000ms (50s). If a larger value is provided,
// it will be automatically limited to 50000ms.
//
// Example:
//
//	cmd.ExecuteCommand("ls -la", WithTimeoutMs(5000))
func WithTimeoutMs(timeoutMs int) CommandOption {
	return func(opts *commandOptions) {
		opts.timeoutMs = timeoutMs
	}
}

// WithCwd sets the working directory for command execution.
// If not set, the command runs in the default session directory.
//
// Example:
//
//	cmd.ExecuteCommand("pwd", WithCwd("/tmp"))
func WithCwd(cwd string) CommandOption {
	return func(opts *commandOptions) {
		opts.cwd = cwd
	}
}

// WithEnvs sets environment variables for command execution.
// These variables are set for the command execution only.
//
// Example:
//
//	cmd.ExecuteCommand("echo $VAR", WithEnvs(map[string]string{"VAR": "value"}))
func WithEnvs(envs map[string]string) CommandOption {
	return func(opts *commandOptions) {
		opts.envs = envs
	}
}

// ExecuteCommand executes a shell command in the session environment.
//
// This method supports both the legacy signature (command string, timeoutMs ...int)
// and the Functional Options pattern for flexible configuration.
//
// Legacy usage (backward compatible):
//   - cmd.ExecuteCommand("ls -la")
//   - cmd.ExecuteCommand("ls -la", 5000)
//
// Functional Options usage (recommended for new code):
//   - cmd.ExecuteCommand("ls -la", WithTimeoutMs(5000))
//   - cmd.ExecuteCommand("pwd", WithCwd("/tmp"), WithEnvs(map[string]string{"VAR": "value"}))
//
// Parameters:
//   - command: The shell command to execute
//   - options: Either an int (timeoutMs in milliseconds) for legacy usage, or
//     CommandOption functions for Functional Options pattern.
//     Maximum allowed timeout is 50000ms (50s). If a larger value is provided,
//     it will be automatically limited to 50000ms
//
// Returns:
//   - *CommandResult: Result containing command output, exit code, stdout,
//     stderr, trace_id, and request ID
//   - error: Error if the operation fails
//
// Example:
//
//	// Default usage
//	cmd.ExecuteCommand("ls")
//
//	// Legacy usage (backward compatible)
//	cmd.ExecuteCommand("ls", 5000)
//
//	// New style with Functional Options
//	cmd.ExecuteCommand("ls", WithTimeoutMs(5000))
//
//	// Combined options
//	cmd.ExecuteCommand("pwd",
//	    WithTimeoutMs(5000),
//	    WithCwd("/tmp"),
//	    WithEnvs(map[string]string{"FOO": "bar"}),
//	)
func (c *Command) ExecuteCommand(command string, options ...interface{}) (*CommandResult, error) {
	// Default configuration
	opts := &commandOptions{
		timeoutMs: 1000, // Default 1 second
		cwd:       "",
		envs:      nil,
	}

	// Handle both old and new styles
	for _, opt := range options {
		switch v := opt.(type) {
		case int: // Old: ExecuteCommand("ls", 5000)
			if v > 0 {
				opts.timeoutMs = v
			}
		case CommandOption: // New: ExecuteCommand("ls", WithTimeoutMs(5000))
			if v != nil {
				v(opts)
			}
		}
	}

	return c.executeCommandInternal(command, opts.timeoutMs, opts.cwd, opts.envs)
}

// Run is an alias of ExecuteCommand.
func (c *Command) Run(command string, options ...interface{}) (*CommandResult, error) {
	return c.ExecuteCommand(command, options...)
}

// Exec is an alias of ExecuteCommand.
func (c *Command) Exec(command string, options ...interface{}) (*CommandResult, error) {
	return c.ExecuteCommand(command, options...)
}

// executeCommandInternal is the internal implementation of command execution
func (c *Command) executeCommandInternal(
	command string,
	timeout int,
	cwd string,
	envs map[string]string,
) (*CommandResult, error) {
	// Limit timeout to maximum 50s (50000ms) as per SDK constraints
	const MAX_TIMEOUT_MS = 50000
	if timeout > MAX_TIMEOUT_MS {
		// Log warning (in production, you might want to use a proper logger)
		// fmt.Printf("Warning: Timeout %dms exceeds maximum allowed %dms. Limiting to %dms.\n", timeout, MAX_TIMEOUT_MS, MAX_TIMEOUT_MS)
		timeout = MAX_TIMEOUT_MS
	}

	// Note: In Go, envs is typed as map[string]string, which enforces type safety at compile time.
	// All keys and values are guaranteed to be strings by the type system.
	// This validation is kept for consistency with other SDKs and documentation purposes.

	// Build request arguments
	args := map[string]interface{}{
		"command":    command,
		"timeout_ms": timeout,
	}
	if cwd != "" {
		args["cwd"] = cwd
	}
	if len(envs) > 0 {
		args["envs"] = envs
	}

	// Use Session's CallMcpTool method
	result, err := c.Session.CallMcpTool("shell", args)
	if err != nil {
		return nil, fmt.Errorf("failed to execute command: %w", err)
	}

	if result.Success {
		// Try to parse the new JSON format response
		var dataJson map[string]interface{}
		if err := json.Unmarshal([]byte(result.Data), &dataJson); err != nil {
			// Fallback to old format if JSON parsing fails
			return &CommandResult{
				ApiResponse: models.ApiResponse{
					RequestID: result.RequestID,
				},
				Success: true,
				Output:  result.Data,
			}, nil
		}

		// Extract fields from new format
		stdout := ""
		if val, ok := dataJson["stdout"].(string); ok {
			stdout = val
		}
		stderr := ""
		if val, ok := dataJson["stderr"].(string); ok {
			stderr = val
		}
		exitCode := 0
		if val, ok := dataJson["exit_code"].(float64); ok {
			exitCode = int(val)
		}
		traceID := ""
		if val, ok := dataJson["traceId"].(string); ok {
			traceID = val
		}

		// Determine success based on exit_code (0 means success)
		success := exitCode == 0

		// For backward compatibility, output should be stdout + stderr
		output := stdout + stderr

		return &CommandResult{
			ApiResponse: models.ApiResponse{
				RequestID: result.RequestID,
			},
			Success:      success,
			Output:       output,
			ExitCode:     exitCode,
			Stdout:       stdout,
			Stderr:       stderr,
			TraceID:      traceID,
			ErrorMessage: result.ErrorMessage,
		}, nil
	} else {
		// Try to parse error message as JSON (in case backend returns JSON in error)
		var errorData map[string]interface{}
		if err := json.Unmarshal([]byte(result.ErrorMessage), &errorData); err == nil {
			// Successfully parsed JSON from error message
			stdout := ""
			if val, ok := errorData["stdout"].(string); ok {
				stdout = val
			}
			stderr := ""
			if val, ok := errorData["stderr"].(string); ok {
				stderr = val
			}
			exitCode := 0
			// Backend may return either "exit_code" or "errorCode", support both
			if val, ok := errorData["exit_code"].(float64); ok {
				exitCode = int(val)
			} else if val, ok := errorData["errorCode"].(float64); ok {
				exitCode = int(val)
			}
			traceID := ""
			if val, ok := errorData["traceId"].(string); ok {
				traceID = val
			}
			// For backward compatibility, output should be stdout + stderr
			output := stdout + stderr

			return &CommandResult{
				ApiResponse: models.ApiResponse{
					RequestID: result.RequestID,
				},
				Success:      false,
				Output:       output,
				ExitCode:     exitCode,
				Stdout:       stdout,
				Stderr:       stderr,
				TraceID:      traceID,
				ErrorMessage: stderr,
			}, nil
		}

		// If parsing fails, return error
		return &CommandResult{
			ApiResponse: models.ApiResponse{
				RequestID: result.RequestID,
			},
			Success:      false,
			Output:       "",
			ErrorMessage: result.ErrorMessage,
		}, nil
	}
}
